### DISCUSSION TITLE
Filter tracer/span in java

### DISCUSSION THREAD

**Discussion by KaeMoNee** [open-telemetry/opentelemetry-java#3167] [Q&A] [answered]
Hi. Is it possible to filter spans by attributes or tracers by names in java without using collector and its configuration for filtering data? (like here [Filter processor](https://github.com/open-telemetry/opentelemetry-collector/blob/main/processor/filterprocessor/README.md))

Thanks in advance!

**Comment by jkwatson** [comment, marked as answer]
I think this could be implemented with a custom exporter wrapper that could filter with whatever logic you wanted. Other than that, we don't have anything built-in like this right now.

**Reply by KaeMoNee** [reply]
Thanks for reply. Can a custom processor wrapper be used as well? Like overriding
```java
  @Override
  public void onEnd(ReadableSpan span) {
    // own filter logic
    // <...>
    delegate.onEnd(span);
  }
```
Or it should be avoided to be done in processor?

**Reply by jkwatson** [reply]
I don't see how you could filter there. You still have to pass a span into your delegate.

**Reply by KaeMoNee** [reply]
I might have written wrong. I thought about not passing it if filter says 'no'.

As BatchSpanProcessor has:
```java
    if (!span.getSpanContext().isSampled()) {
      return;
    }
    worker.addSpan(span);
```

I thought about:
```java
  @Override
  public void onEnd(ReadableSpan span) {
    if (!filter.canSend(span)) {
      return;
    }
    delegate.onEnd(span);
  }
```

**Reply by jkwatson** [reply]
ah, I see. Yes, that is definitely a possibility. You'll need to worry about the possibility of removing a span in the middle of a trace, which could break the trace in the middle if the other spans are included.

**Reply by KaeMoNee** [reply]
Thanks a lot. I will think about `the possibility of removing a span in the middle of a trace`, but at least now I know how to implement filter for my case.

**Comment by Oberon00** [comment]
This ought to be possible with a sampler but sadly isn't: open-telemetry/opentelemetry-specification#1588

**Reply by KaeMoNee** [reply]
I tried to use a sampler for it, but I had no access to attributes for given span (as I understood, span was not created yet, only name, because Span.current() would return me parent span). If span is not yet created (? correct me if I'm wrong) then I still cannot have any information about attributes inside.

Also, I could not find any way to receive attributes from a span (maybe only by checking if this span implements ReadableSpan to use toSpanData() method).

**Reply by Oberon00** [reply]
The sampler receives the attributes as parameter in `ShouldSample`, see here: https://github.com/open-telemetry/opentelemetry-java/blob/v1.1.0/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/samplers/Sampler.java#L110

**Reply by KaeMoNee** [reply]
In my case I had empty attributes (using 0.16.0 version). This happens (probably) because I put attributes after `.startSpan()`.

**Reply by Oberon00** [reply]
Yes, the sampler only has access to attributes set via the SpanBuilder. That's why the specification recommends setting attributes before starting the span if possible, see https://github.com/open-telemetry/opentelemetry-specification/blob/v1.2.0/specification/trace/api.md#span-creation:

> Whenever possible, users SHOULD set any already known attributes at span creation instead of calling SetAttribute later.
